packages/Search/src/details/DTK_DetailsDistributor.hpp

```diff
@@ -0,0 +1,252 @@
+                                  outdegrees, notNullPtr( destinations.data() ),
+                                  MPI_UNWEIGHTED );
+
+        _src_counts.resize( indegrees );
+        MPI_Neighbor_alltoall( _dest_counts.data(), 1, MPI_INT,
+                               _src_counts.data(), 1, MPI_INT,
+                               _comm_dist_graph );
+
+        _src_offsets.resize( indegrees + 1 );
+        // exclusive scan
+        _src_offsets[0] = 0;
+        for ( int i = 0; i < indegrees; ++i )
+            _src_offsets[i + 1] = _src_offsets[i] + _src_counts[i];
+
+#if defined( REORDER_RECV )
+        int comm_size = -1;
+        MPI_Comm_size( _comm, &comm_size );
+        _permute_recv.resize( _src_offsets.back() );
+        int offset = 0;
+        for ( int i = 0; i < comm_size; ++i )
+        {
+            auto const it = std::find( sources.begin(), sources.end(), i );
+            if ( it != sources.end() )
+            {
+                int const j = std::distance( sources.begin(), it );
+                std::iota( &_permute_recv[offset],
+                           &_permute_recv[offset] + _src_counts[j],
+                           _src_offsets[j] );
+                offset += _src_counts[j];
+            }
+        }
+        DTK_ENSURE( offset == static_cast<int>( _permute_recv.size() ) );
+#endif
+
+        return _src_offsets.back();
+    }
+    template <typename View>
+    void doPostsAndWaits( typename View::const_type const &exports,
+                          size_t num_packets, View const &imports ) const
+    {
+        DTK_REQUIRE( num_packets * _src_offsets.back() == imports.size() );
+        DTK_REQUIRE( num_packets * _dest_offsets.back() == exports.size() );
+
+        using ValueType = typename View::value_type;
+        static_assert( View::rank == 1, "" );
+        static_assert(
+            Kokkos::Impl::MemorySpaceAccess<typename View::memory_space,
+                                            Kokkos::HostSpace>::accessible,
+            "" );
+
+        std::vector<int> dest_counts = _dest_counts;
+        std::vector<int> dest_offsets = _dest_offsets;
+        std::vector<int> src_counts = _src_counts;
+        std::vector<int> src_offsets = _src_offsets;
+        for ( auto pv :
+              {&dest_counts, &dest_offsets, &src_counts, &src_offsets} )
+            for ( auto &x : *pv )
+                x *= num_packets * sizeof( ValueType );
+
+        std::vector<ValueType> dest_buffer( exports.size() );
+        std::vector<ValueType> src_buffer( imports.size() );
+
+        // TODO
+        // * apply permutation on the device in a parallel for
+        // * switch to MPI with CUDA support (do not copy to host)
+        for ( int i = 0; i < _dest_offsets.back(); ++i )
+            std::copy( &exports[num_packets * i],
+                       &exports[num_packets * i] + num_packets,
+                       &dest_buffer[num_packets * _permute[i]] );
+
+        MPI_Neighbor_alltoallv(
```

Comment:
Yeah this is so much easier now that I see all of this. We will definitely want to do some performance tests in Cabana to see which way performs better. This way is certainly cleaner and more maintainable.